sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java

```diff
@@ -815,13 +816,169 @@ public void validate(PipelineOptions options) {
+
+                          if (tableReadOptionsBuilder != null) {
+                            requestBuilder.setReadOptions(tableReadOptionsBuilder);
+                          }
+                        }
+
+                        ParallelRead.CreateSessionRequest request = requestBuilder.build();
+                        ParallelRead.Session session = getBigQueryServices()
+                            .getTableReadService(c.getPipelineOptions().as(BigQueryOptions.class))
+                            .createSession(request);
+
+                        c.output(sessionTag, session);
+
+                        for (ParallelRead.ReadLocation initialReadLocation :
+                            session.getInitialReadLocationsList()) {
+                          c.output(initialReadLocation);
+                        }
+                      }
+                    })
+                    .withOutputTags(readLocationTag,
+                        TupleTagList.of(tableReferenceTag).and(sessionTag)));
+
+        tuple.get(sessionTag).setCoder(ProtoCoder.of(ParallelRead.Session.class));
+        tuple.get(readLocationTag).setCoder(ProtoCoder.of(ParallelRead.ReadLocation.class));
+        tuple.get(tableReferenceTag).setCoder(StringUtf8Coder.of());
+
+        final PCollectionView<ParallelRead.Session> sessionView =
+            tuple.get(sessionTag).apply("ViewSessionId", View.asSingleton());
+
+        PCollection<T> rows = tuple
+            .get(readLocationTag)
+            .apply(Reshuffle.viaRandomKey())
+            .apply(
+                "ReadFromTemporaryTable",
+                ParDo.of(
+                    new DoFn<ParallelRead.ReadLocation, T>() {
+                      @ProcessElement
+                      public void processElement(ProcessContext c) throws Exception {
+
+                        BigQueryParallelReadStreamSource<T> source =
+                            BigQueryParallelReadStreamSource.create(
+                                getBigQueryServices(),
+                                getRowProtoParseFn(),
+                                coder,
+                                getReadSessionOptions(),
+                                c.sideInput(sessionView),
+                                c.element(),
+                                null);
+
+                        BoundedSource.BoundedReader<T> reader =
+                            source.createReader(c.getPipelineOptions());
+                        for (boolean more = reader.start(); more; more = reader.advance()) {
+                          c.output(reader.getCurrent());
+                        }
+                      }
+                    })
+                    .withSideInputs(sessionView));
+
+        rows.setCoder(coder);
+
+        final PCollectionView<String> tableReferenceView =
+            tuple.get(tableReferenceTag).apply("ViewTemporaryTable", View.asSingleton());
+
+        rows.apply("WaitForCompletion", Count.globally())
```

Comment:
I'm not sure that this provides enough guarantees. The fact that Count.globally() has returned a result doesn't guarantee that all other transforms using "rows" have finished - "ReadFromTemporaryTable" may run again even after DeleteTemporaryTable runs. Is it possible to instead set a TTL on the temporary table?

Reply:
After offline discussion, I have converted this to use PassThroughThenCleanup.